File: clients/python/src/novamem/_transport.py
"""One function makes every request, so the classification of failures is
written exactly once."""

from __future__ import annotations

import http.client
import json
import urllib.error
import urllib.parse
import urllib.request
from typing import Any

from ._errors import ConfigError, NotFoundError, NovamemError, UnavailableError

#: Bounds a single call when no timeout is given. Sized for the slowest
#: operation: capture embeds the content server-side before it answers.
DEFAULT_TIMEOUT = 15.0

#: Responses larger than this are rejected rather than buffered: a rogue
#: endpoint must not be able to make the caller allocate without bound.
MAX_RESPONSE_BYTES = 8 << 20


class Transport:
    def __init__(
        self,
        base_url: str,
        token: str,
        *,
        timeout: float = DEFAULT_TIMEOUT,
        opener: urllib.request.OpenerDirector | None = None,
    ) -> None:
        base = (base_url or "").strip().rstrip("/")
        parts = urllib.parse.urlsplit(base)
        if parts.scheme not in ("http", "https") or not parts.netloc:
            # Names the field, never the value: a token pasted into the
            # wrong argument must not end up in a log line.
            raise ConfigError("novamem: base_url is not an absolute http(s) URL")
        if not (token or "").strip():
            raise ConfigError("novamem: token is required")
        self.base_url = base
        self._token = token
        self.timeout = timeout if timeout and timeout > 0 else DEFAULT_TIMEOUT
        self._opener = opener or urllib.request.build_opener(_SameOriginAuth())

    def __repr__(self) -> str:
        return f"Transport(base_url={self.base_url!r}, token=[redacted], timeout={self.timeout})"

    def _redact(self, s: str) -> str:
        return s.replace(self._token, "[redacted]")

    def call(
        self,
        op: str,
        method: str,
        path: str,
        body: Any = None,
        *,
        query: dict[str, Any] | None = None,
        expect_body: bool = True,
    ) -> Any:
        """Performs one request and returns the decoded JSON body (or None
        when ``expect_body`` is False). Raises NovamemError on anything else.

        The timeout bounds each socket operation (connect, and every read),
        which is what the standard library offers; a server that answers a
        byte at a time can stretch a call beyond it.
        """
        url = self.base_url + path
        q = {k: v for k, v in (query or {}).items() if v is not None and v != ""}
        if q:
            url += "?" + urllib.parse.urlencode({k: _qs(v) for k, v in q.items()})
        headers = {
            "Accept": "application/json",
            "Authorization": "Bearer " + self._token,
        }
        data = None
        if body is not None:
            data = json.dumps(body, separators=(",", ":")).encode()
            headers["Content-Type"] = "application/json"
        req = urllib.request.Request(url, data=data, method=method, headers=headers)

        try:
            with self._opener.open(req, timeout=self.timeout) as resp:
                status, raw = resp.status, resp.read(MAX_RESPONSE_BYTES + 1)
        except urllib.error.HTTPError as e:
            try:
                raw = e.read(MAX_RESPONSE_BYTES + 1)
            except (OSError, http.client.HTTPException):
                raw = b""
            finally:
                e.close()
            raise self._http_error(op, e.code, raw) from None
        except urllib.error.URLError as e:
            if isinstance(e.reason, TimeoutError):
                raise UnavailableError(op, "timed out", retryable=True) from None
            raise UnavailableError(
                op, self._redact(f"unreachable: {e.reason}"), retryable=True
            ) from None
        except TimeoutError:
            raise UnavailableError(op, "timed out", retryable=True) from None
        except (OSError, http.client.HTTPException) as e:
            # Reset connection, a server that closed without answering, a
            # truncated read: the host could not be consulted.
            raise UnavailableError(
                op, self._redact(f"unreachable: {e!r}"), retryable=True
            ) from None

        if len(raw) > MAX_RESPONSE_BYTES:
            raise UnavailableError(
                op, "response body exceeds 8 MiB", status_code=status
            )
        if not expect_body:
            return None
        if not raw.strip():
            # A 2xx with no body is not the contract. Decoding it into a
            # default would tell a forget caller the delete happened.
            raise UnavailableError(op, "empty response body", status_code=status)
        try:
            return json.loads(raw)
        except ValueError:
            # In practice a proxy's HTML error page or a truncated response:
            # we never reached a working novamem. Not retryable — the same
            # request parses the same way.
            raise UnavailableError(
                op, "malformed response body", status_code=status
            ) from None

    def _http_error(self, op: str, status: int, raw: bytes) -> NovamemError:
        message, code = "", ""
        try:
            payload = json.loads(raw)
            if isinstance(payload, dict) and payload.get("error"):
                message = str(payload["error"])
                code = str(payload.get("code") or "")
        except ValueError:
            pass
        if not message:
            text = raw.decode("utf-8", "replace").strip()
            message = text[:256] + "…" if len(text) > 256 else text
        # The server's message and code are quoted verbatim, and a server that
        # echoes the credential back ("bad token nm_…") would otherwise
        # launder it into this client's logs.
        message = self._redact(message)
        code = self._redact(code)
        if status >= 500 or status == 429:
            return UnavailableError(
                op, message, status_code=status, code=code, retryable=True
            )
        if status == 404:
            return NotFoundError(op, message, status_code=status, code=code)
        return NovamemError(op, message, status_code=status, code=code)


class _SameOriginAuth(urllib.request.HTTPRedirectHandler):
    """Follows redirects, but never carries the bearer to another origin.

    urllib's default handler copies every header onto the redirected
    request, so a redirect to a different host would hand it the token. Go's
    http.Client drops Authorization on a cross-origin redirect; so does this.
    """

    def redirect_request(self, req, fp, code, msg, headers, newurl):  # type: ignore[no-untyped-def]
        new = super().redirect_request(req, fp, code, msg, headers, newurl)
        if new is not None and _origin(new.full_url) != _origin(req.full_url):
            new.remove_header("Authorization")
        return new


def _origin(url: str) -> tuple[str, str]:
    p = urllib.parse.urlsplit(url)
    return (p.scheme, p.netloc.lower())


def _qs(v: Any) -> str:
    if isinstance(v, bool):
        return "true" if v else "false"
    return str(v)

File: clients/swift/Sources/Novamem/Transport.swift
import Foundation
#if canImport(FoundationNetworking)
    import FoundationNetworking
#endif

/// Responses larger than this are rejected rather than kept.
public let maxResponseBytes = 8 << 20

/// Follows redirects, but never carries the bearer to another origin: Go's
/// http.Client drops Authorization on a cross-origin hop, and so does this.
final class SameOriginAuth: NSObject, URLSessionTaskDelegate, @unchecked Sendable {
    func urlSession(_: URLSession, task: URLSessionTask, willPerformHTTPRedirection _: HTTPURLResponse,
                    newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void)
    {
        // URLSession builds the redirected request without Authorization, so
        // a same-origin hop has to carry it over explicitly; a cross-origin
        // one must not.
        var next = request
        let from = task.currentRequest ?? task.originalRequest
        if let from, let a = from.url, let b = request.url, Self.sameOrigin(a, b) {
            next.setValue(from.value(forHTTPHeaderField: "Authorization"), forHTTPHeaderField: "Authorization")
        } else {
            next.setValue(nil, forHTTPHeaderField: "Authorization")
        }
        completionHandler(next)
    }

    static func sameOrigin(_ a: URL, _ b: URL) -> Bool {
        func port(_ u: URL) -> Int {
            u.port ?? (u.scheme == "https" ? 443 : 80)
        }
        return a.scheme == b.scheme && a.host?.lowercased() == b.host?.lowercased() && port(a) == port(b)
    }
}

/// One function makes every request, so failures are classified once.
final class Transport: @unchecked Sendable {
    let config: Config
    private let session: URLSession

    init(_ config: Config) {
        self.config = config
        let c = URLSessionConfiguration.ephemeral
        c.timeoutIntervalForRequest = config.timeout
        c.timeoutIntervalForResource = config.timeout
        session = URLSession(configuration: c, delegate: SameOriginAuth(), delegateQueue: nil)
    }

    deinit { session.finishTasksAndInvalidate() }

    private func redact(_ s: String) -> String {
        s.replacingOccurrences(of: config.token, with: "[redacted]")
    }

    /// Performs one request; returns its status and JSON bytes (nil when
    /// `expectBody` is false) or throws NovamemError / CancellationError.
    func call(_ op: String, _ method: String, _ path: String, body: Data? = nil,
              query: [String: String?] = [:], expectBody: Bool = true) async throws -> (status: Int, body: Data?)
    {
        var comps = URLComponents(string: config.baseURL + path)!
        let q = query.compactMap { k, v in v.flatMap { $0.isEmpty ? nil : URLQueryItem(name: k, value: $0) } }
        if !q.isEmpty {
            comps.queryItems = q.sorted { $0.name < $1.name }
        }
        var req = URLRequest(url: comps.url!)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.setValue("Bearer \(config.token)", forHTTPHeaderField: "Authorization")
        if let body {
            req.httpBody = body
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        let (status, raw) = try await exchange(op, req)
        return try (status, decode(op, status, raw, expectBody))
    }

    private func exchange(_ op: String, _ req: URLRequest) async throws -> (Int, Data) {
        let box = TaskBox()
        return try await withTaskCancellationHandler {
            try await withCheckedThrowingContinuation { (k: CheckedContinuation<(Int, Data), Error>) in
                let task = session.dataTask(with: req) { data, response, error in
                    if let error {
                        k.resume(throwing: self.transportError(op, error))
                        return
                    }
                    k.resume(returning: ((response as? HTTPURLResponse)?.statusCode ?? 0, data ?? Data()))
                }
                box.set(task)
                task.resume()
            }
        } onCancel: {
            box.cancel()
        }
    }

    private func transportError(_ op: String, _ error: Error) -> Error {
        let e = error as NSError
        if e.domain == NSURLErrorDomain, e.code == NSURLErrorCancelled {
            return CancellationError()
        }
        if e.domain == NSURLErrorDomain, e.code == NSURLErrorTimedOut {
            return NovamemError(op: op, message: "timed out", unavailable: true, retryable: true)
        }
        // Refused dial, DNS, reset, TLS: the host could not be consulted.
        return NovamemError(op: op, message: redact("unreachable: \(e.localizedDescription)"), unavailable: true, retryable: true)
    }

    private func decode(_ op: String, _ status: Int, _ raw: Data, _ expectBody: Bool) throws -> Data? {
        if raw.count > maxResponseBytes {
            throw NovamemError(op: op, message: "response body exceeds 8 MiB", statusCode: status, unavailable: true)
        }
        guard (200 ..< 300).contains(status) else { throw httpError(op, status, raw) }
        if !expectBody {
            return nil
        }
        // A 2xx with no body is not the contract: decoding it into a default
        // would tell a forget caller the delete happened.
        if String(decoding: raw, as: UTF8.self).trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            throw NovamemError(op: op, message: "empty response body", statusCode: status, unavailable: true)
        }
        // In practice a proxy's HTML error page: we never reached a working
        // novamem. Not retryable — the same request parses the same way.
        guard (try? JSONSerialization.jsonObject(with: raw, options: [.fragmentsAllowed])) != nil else {
            throw NovamemError(op: op, message: "malformed response body", statusCode: status, unavailable: true)
        }
        return raw
    }

    private func httpError(_ op: String, _ status: Int, _ raw: Data) -> NovamemError {
        var message = ""
        var code = ""
        if let o = (try? JSONSerialization.jsonObject(with: raw)) as? [String: Any], let e = o["error"] as? String, !e.isEmpty {
            message = e
            code = (o["code"] as? String) ?? ""
        }
        if message.isEmpty {
            let text = String(decoding: raw, as: UTF8.self).trimmingCharacters(in: .whitespacesAndNewlines)
            message = text.count > 256 ? String(text.prefix(256)) + "…" : text
        }
        // The server's message and code are quoted verbatim; a server echoing
        // the credential back would otherwise launder it into the logs.
        let unavailable = status >= 500 || status == 429
        return NovamemError(op: op, message: redact(message), statusCode: status, code: redact(code),
                            unavailable: unavailable, retryable: unavailable)
    }
}

/// Holds a task so a cancellation arriving before it exists still lands.
private final class TaskBox: @unchecked Sendable {
    private let lock = NSLock()
    private var task: URLSessionTask?
    private var cancelled = false

    func set(_ t: URLSessionTask) {
        lock.lock()
        task = t
        let c = cancelled
        lock.unlock()
        if c {
            t.cancel()
        }
    }

    func cancel() {
        lock.lock()
        cancelled = true
        let t = task
        lock.unlock()
        t?.cancel()
    }
}
